client/block-builder/src/lib.rs

```diff
@@ -40,9 +45,42 @@ use sp_runtime::{
 	Digest,
 };
 
+use sc_client_api::backend;
+use sc_executor::RuntimeVersionOf;
 pub use sp_block_builder::BlockBuilder as BlockBuilderApi;
+use sp_core::storage::{well_known_keys, StateVersion, Storage};
 
-use sc_client_api::backend;
+/// Return the state version given the genesis storage and executor.
+pub fn resolve_state_version_from_wasm<E>(
+	storage: &Storage,
+	executor: &E,
+) -> sp_blockchain::Result<StateVersion>
+where
+	E: RuntimeVersionOf,
+{
+	if let Some(wasm) = storage.top.get(well_known_keys::CODE) {
+		let mut ext = sp_state_machine::BasicExternalities::new_empty(); // just to read runtime version.
+
+		let code_fetcher = sp_core::traits::WrappedRuntimeCode(wasm.as_slice().into());
+		let runtime_code = sp_core::traits::RuntimeCode {
+			code_fetcher: &code_fetcher,
+			heap_pages: None,
+			hash: {
+				use std::hash::{Hash, Hasher};
+				let mut state = DefaultHasher::new();
+				wasm.hash(&mut state);
+				state.finish().to_le_bytes().to_vec()
+			},
+		};
+		let runtime_version = RuntimeVersionOf::runtime_version(executor, &mut ext, &runtime_code)
```

Comment:
Have you considered using [ReadRuntimeVersion](https://paritytech.github.io/substrate/master/sp_core/traits/trait.ReadRuntimeVersion.html) trait + a manual decode instead? I think that could probably work? And it wouldn't require all of the fiddling with `WrappedRuntimeCode` and `RuntimeCode`.